pkg/descheduler/strategies/topologyspreadconstraint.go

```diff
@@ -0,0 +1,242 @@
+	// if diff > maxSkew, add this pod in the current bucket for eviction
+
+	evictable := podEvictor.Evictable()
+
+	namespacedTopologySpreadConstrainPods := make(map[string][]v1.Pod)
+	// First record all of the constraints by namespace
+	namespaces, err := client.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		klog.ErrorS(err, "couldn't list namespaces")
+		return
+	}
+	podsForEviction := make(map[*v1.Pod]struct{})
+	// 1. for each namespace...
+	for _, namespace := range namespaces.Items {
+		namespacePods, err := client.CoreV1().Pods(namespace.Name).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			klog.ErrorS(err, "couldn't list pods in namespace", "namespace", namespace)
+			return
+		}
+
+		// ...where there is a topology constraint
+		for _, pod := range namespacePods.Items {
+			if pod.Spec.TopologySpreadConstraints != nil {
+				namespacedTopologySpreadConstrainPods[pod.Namespace] = append(namespacedTopologySpreadConstrainPods[pod.Namespace], pod)
+			}
+		}
+		if len(namespacedTopologySpreadConstrainPods[namespace.Name]) == 0 {
+			continue
+		}
+
+		// 2. for each topologySpreadConstraint in that namespace
+		for _, constrainedPod := range namespacedTopologySpreadConstrainPods[namespace.Name] {
+			// find pods to evict one constraint at a time
+			for _, constraint := range constrainedPod.Spec.TopologySpreadConstraints {
+				constraintTopologies := make(map[topologyPair][]*v1.Pod)
+
+				// 3. for each evictable pod in that namespace
+				// (this loop is where we count the number of pods per topologyValue that match this constraint's selector)
+				var sumPods float64
+				for _, pod := range namespacePods.Items {
+					if !evictable.IsEvictable(&pod) {
+						continue
+					}
+
+					// 4. if the pod matches this TopologySpreadConstraint LabelSelector
+					s, err := metav1.LabelSelectorAsSelector(constraint.LabelSelector)
+					if err != nil {
+						klog.ErrorS(err, "couldn't parse label selector as selector", "selector", constraint.LabelSelector)
+					}
+					if !s.Matches(labels.Set(pod.Labels)) {
+						continue
+					}
+
+					// 5. If the pod's node matches this constraint's topologyKey, create a topoPair and add the pod
+					node, err := client.CoreV1().Nodes().Get(ctx, pod.Spec.NodeName, metav1.GetOptions{})
+					if err != nil {
+						klog.ErrorS(err, "couldn't get node", "node", pod.Spec.NodeName)
+					}
+					nodeValue, ok := node.Labels[constraint.TopologyKey]
+					if !ok {
+						continue
+					}
+					// 6. create a topoPair with key as this TopologySpreadConstraint
+					topoPair := topologyPair{key: constraint.TopologyKey, value: nodeValue}
+					// 7. add the pod with key as this topoPair
+					constraintTopologies[topoPair] = append(constraintTopologies[topoPair], &pod)
+					sumPods++
+				}
+
+				// To actually determine how many pods need to be moved, we sort the topologies in ascending length
```

Comment:
I added what I think is a better algorithm for determining how many pods to evict from topologies here (in https://github.com/kubernetes-sigs/descheduler/pull/413/commits/a9976c1b005ed1e1d8c1a532a088fd034f1c86a2)

I put a lot of description in the comments, but basically what I'm doing is splitting the topologies into 2 lists, balanced around the "ideal average" (ie, the size if all pods were perfectly spread). Mathematically, it seems like all topologies need to fall within `maxSkew` of this average. So I built around those that are "borrowing" from the average to be above it, and those that are "lending" from the average to be below it.

So I compare the smallest topology to the largest topology (these are the most likely to be skewed, and the largest topology will have the most "borrowed" that it is able to pay back). If the largest topology has enough to satisfy the `maxSkew` of the smallest topology, we move on to check the same with the next largest "small" topology. If it does not, we move on to check with the next *smallest* "large" topology.

I've tried this with a couple manual examples but would like some help verifying that my assumptions are right. This would run in linear time and simplify the data structures needed to keep track of how many pods have been moved.